src/dom-to-image-more.js

```diff
@@ -1199,13 +1199,31 @@
                 sandbox.style.visibility = 'hidden';
                 sandbox.style.position = 'fixed';
                 document.body.appendChild(sandbox);
-                // Ensure that the iframe is rendered in standard mode
+                sandbox.id = sandbox.contentDocument.title = 'domtoimage-sandbox-' + util.uid();
+                // Ensure the iframe's character rendering matches the document (with UTF-8 fallback).
                 const charset = document.createElement('meta');
                 charset.setAttribute('charset', document.characterSet || 'UTF-8');
                 sandbox.contentDocument.head.appendChild(charset);
-                sandbox.contentDocument.title = 'sandbox';
+                // Ensure the document is rendered in standard mode by coercing doctype.
+                // If the parent document has a Trusted Types policy, fallback to quirks mode.
+                let isTrustedTypesRequired;
+                try {
+                    const request = new XMLHttpRequest();
+                    request.open('HEAD', document.location.href, false);
+                    request.send();
+                    const csp = request.getResponseHeader('content-security-policy');
+                    isTrustedTypesRequired = csp.indexOf('require-trusted-types-for') !== -1;
```

Comment:
To check for Trusted Type API error, instead of creating a network request, what if we try the original `sandbox.contentWindow.document.write` in the `try` block? Then in the `catch` block, we execute the fallback code.

Another idea - would `documeny.write` succeed if we used `DOMPurity.sanitize` first? See an example [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/require-trusted-types-for#examples)

Reply:
One more thing - using XMLHttpRequest is deprecated. I get the following warning in my console - "Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help http://xhr.spec.whatwg.org/"

Reply:
Deep into this rabbit-hole... stumbled upon this comment https://github.com/w3c/trusted-types/blob/main/src/trustedtypes.js#L337 which worried me a bit, but I think that's only Edge 12-18 (e.g. the not-Chromium-based version) https://caniuse.com/iframe-srcdoc 

Noting this here so I don't head down the same rabbit hole.

Reply:
Looks like we could just look to see if there's a CSP on the currently executing script like https://github.com/w3c/trusted-types/blob/main/src/polyfill/full.js#L27

Reply:
😄 seeing as we are in the rabbit hole ...

Try-finally doesn't work for some strange reason. However we can actually use the Trusted Types API to check if a policy is needed here - [spec](https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getpropertytype)

Reply:
😄 seeing as we are in the rabbit hole ...

try-finally doesn't work for some strange reason. However we can actually use the Trusted Types API to check if a policy is needed here - [spec](https://w3c.github.io/trusted-types/dist/spec/#dom-trustedtypepolicyfactory-getpropertytype)

Reply:
Trying to get somewhere , how does this look?
```javascript
        function ensureSandboxWindow() {
            if (!sandbox) {
                // figure out how this document is defined (doctype and charset)
                const charsetToUse = document.characterSet || 'UTF-8';               
                const docType = document.doctype;
                const docTypeDeclaration = docType
                    ? `<!DOCTYPE ${docType.name} ${docType.publicId} ${docType.systemId}`.trim() + '>'
                    : '';

                // Create a hidden sandbox <iframe> element within we can create default HTML elements and query their
                // computed styles. Elements must be rendered in order to query their computed styles. The <iframe> won't
                // render at all with `display: none`, so we have to use `visibility: hidden` with `position: fixed`.
                sandbox = document.createElement('iframe');
                sandbox.id = 'domtoimage-sandbox-' + util.uid();
                sandbox.style.visibility = 'hidden';
                sandbox.style.position = 'fixed';
                document.body.appendChild(sandbox);

                tryTechniques(sandbox, docTypeDeclaration, charsetToUse, 'domtoimage-sandbox');
            }

            return sandbox.contentWindow;
        }

        function tryTechniques(sandbox, doctype, charset, title) {
            // try the good old-fashioned document write with all the correct attributes set
            try {
                sandbox.contentWindow.document.write(`${doctype}<html><head><meta charset='${charset}'><title>${title}</title></head><body></body></html>`);
                return sandbox.contentWindow;
            }
            catch (_) {
                // swallow exception and fall through to next technique
            }

            const metaCharset = document.createElement('meta');
            metaCharset.setAttribute('charset', charset);

            // let's attempt it using srcdoc, so we can still set the doctype and charset
            try {
                const sandboxDocument = document.implementation.createHTMLDocument(title);
                sandboxDocument.head.appendChild(metaCharset);
                const sandboxHTML = doctype + sandboxDocument.documentElement.outerHTML;
                sandbox.setAttribute('srcdoc', sandboxHTML);
                return sandbox.contentWindow;
            } catch (_) {
                // swallow exception and fall through to the simplest path
            }

            // let's attempt it using srcdoc... here we're not able to set the doctype
            sandbox.contentDocument.head.appendChild(metaCharset);
            sandbox.contentDocument.title = title;
            return sandbox.contentWindow;
        }

```

Reply:
I've pushed that as `quirks-regression` branch https://github.com/1904labs/dom-to-image-more/tree/quirks-regression

Can you validate @joswhite as I don't have a fiddle or unit test that shows this.

Reply:
```diff
 - let's attempt it using srcdoc
 + let's attempt it using contentDocument
```

Reply:
Bummer...

<img width="339" alt="image" src="https://user-images.githubusercontent.com/874059/215984857-e5f01e94-cd6e-4bf7-934e-82ebc8d57b61.png">

doesn't look much like

<img width="239" alt="image" src="https://user-images.githubusercontent.com/874059/215985116-3778525d-3674-44ee-8c4e-9747565e859a.png">


```console
await domtoimage.toSvg(document.body)
VM183:1221 
        
       This document requires 'TrustedHTML' assignment.
tryTechniques @ VM183:1221
ensureSandboxWindow @ VM183:1212
getDefaultStyle @ VM183:1151
copyUserComputedStyleFast @ VM183:1071
copyStyle @ VM183:367
cloneStyle @ VM183:342
Promise.then (async)
processClone @ VM183:333
(anonymous) @ VM183:266
Promise.then (async)
cloneNode @ VM183:265
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:76
Promise.then (async)
toSvg @ VM183:75
(anonymous) @ VM188:1
VM183:1236 
        
       This document requires 'TrustedHTML' assignment.
tryTechniques @ VM183:1236
ensureSandboxWindow @ VM183:1212
getDefaultStyle @ VM183:1151
copyUserComputedStyleFast @ VM183:1071
copyStyle @ VM183:367
cloneStyle @ VM183:342
Promise.then (async)
processClone @ VM183:333
(anonymous) @ VM183:266
Promise.then (async)
cloneNode @ VM183:265
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:294
Promise.then (async)
(anonymous) @ VM183:293
cloneChildren @ VM183:292
(anonymous) @ VM183:263
Promise.then (async)
cloneNode @ VM183:262
(anonymous) @ VM183:76
Promise.then (async)
toSvg @ VM183:75
(anonymous) @ VM188:1
callout:1 
        
       Access to XMLHttpRequest at 'https://www.google.com/images/hpp/Chrome_Owned_96x96.png' from origin 'https://ogs.google.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
VM183:764 
        
       cannot fetch resource: https://www.google.com/images/hpp/Chrome_Owned_96x96.png, status: 0
fail @ VM183:764
done @ VM183:738
XMLHttpRequest.send (async)
(anonymous) @ VM183:719
getAndEncode @ VM183:707
Promise.then (async)
inline @ VM183:997
(anonymous) @ VM183:1016
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
(anonymous) @ VM183:1020
(anonymous) @ VM183:1019
Promise.then (async)
inlineAll @ VM183:1014
inlineImages @ VM183:483
Promise.then (async)
toSvg @ VM183:79
(anonymous) @ VM188:1
VM183:719 
        
      
        
        
      
        
      
       
        
       GET https://www.google.com/images/hpp/Chrome_Owned_96x96.png net::ERR_FAILED 200
```

Reply:
The Trusted Types API fn only tells us if we _could_ be hitting the problem, it isn't a requirement check.  
Looks like async XHR with HEAD method is the only sane check we could use here.

Reply:
The API `detectPolicy` snippet doesn't handle trusted type policy that is set as response header on server-side. :(  
Async XHR with HEAD method it is..

Reply:
Sorry, this has been so painful. Thanks for the patience.

Reply:
> I've pushed that as `quirks-regression` branch https://github.com/1904labs/dom-to-image-more/tree/quirks-regression and it runs all the unit tests correctly
> 
> Can you validate @joswhite as I don't have a fiddle or unit test that shows this.

I haven't been able to prove that the old code actually activated Quirks mode. For other `<iframe>` instances without DOCTYPE, I get this. Not sure why I couldn't get the old code to show me the same warning. It might depend on the environment or browser as well.

![image](https://user-images.githubusercontent.com/7979277/216207289-94c957cb-ba24-4a42-b494-231d9ede3f23.png)

Anyways, I didn't see any obvious issues. Thanks so much for your work on this!